File: scripts/prepare-electron-builder-cache.mjs
// Pre-populate electron-builder caches for Windows release builds.
//
// Why is this needed?  electron-builder downloads 3 zip archives at build
// time and extracts them with 7zip.  Two of them contain POSIX symlinks:
//
//   winCodeSign-2.6.0.7z  — darwin .dylib symlinks
//   nsis-3.0.4.1.7z       — (harmless, but still needs extraction)
//   nsis-resources-3.4.0.7z — StdUtils, Nsis7z plugins (missing from nsis)
//
// Windows users lack SeCreateSymbolicLinkPrivilege by default, so 7zip
// fails with "Access is denied" on symbolic links.  This script downloads
// all three archives and extracts them with -snl (skip symbolic links),
// pre-staging the cache so electron-builder never has to.
//
// Usage:
//   node scripts/prepare-electron-builder-cache.mjs
//
// This is safe to run repeatedly (idempotent — skips already-populated dirs).

import { mkdirSync, existsSync, rmSync } from 'node:fs';
import { join, dirname } from 'node:path';
import { tmpdir, homedir } from 'node:os';
import { spawnSync } from 'child_process';
import { fileURLToPath } from 'url';

const __dirname = dirname(fileURLToPath(import.meta.url));
const root = join(__dirname, '..');

const CACHE = join(homedir(), 'AppData', 'Local', 'electron-builder', 'Cache');
const SEVEN_ZIP = join(root, 'node_modules', '7zip-bin', 'win', 'x64', '7za.exe');

const ARCHIVES = [
  {
    name: 'winCodeSign',
    version: 'winCodeSign-2.6.0',
    url: 'https://github.com/electron-userland/electron-builder-binaries/releases/download/winCodeSign-2.6.0/winCodeSign-2.6.0.7z',
    needsSnl: true, // macOS .dylib symlinks
  },
  {
    name: 'nsis',
    version: 'nsis-3.0.4.1',
    url: 'https://github.com/electron-userland/electron-builder-binaries/releases/download/nsis-3.0.4.1/nsis-3.0.4.1.7z',
    needsSnl: false,
  },
  {
    name: 'nsis-resources',
    version: 'nsis-3.0.4.1', // extracted into the same nsis-3.0.4.1 dir
    url: 'https://github.com/electron-userland/electron-builder-binaries/releases/download/nsis-resources-3.4.0/nsis-resources-3.4.0.7z',
    needsSnl: false,
  },
];

async function download(url, dest) {
  console.log(`  downloading ${url}`);
  const res = await fetch(url);
  if (!res.ok || !res.body) {
    throw new Error(`HTTP ${res.status}`);
  }
  const ab = await res.arrayBuffer();
  const buf = Buffer.from(ab);
  const { writeFileSync } = await import('node:fs');
  writeFileSync(dest, buf);
  console.log(`  saved ${(buf.length / 1048576).toFixed(1)} MiB`);
}

function extract(archivePath, destDir, needsSnl) {
  const args = ['x', archivePath, `-o${destDir}`, '-y'];
  if (needsSnl) args.push('-snl'); // skip symbolic links
  const r = spawnSync(SEVEN_ZIP, args, { stdio: 'inherit' });
  if (r.status !== 0) {
    throw new Error(`7za extract failed (exit ${r.status})`);
  }
}

async function main() {
  mkdirSync(CACHE, { recursive: true });

  for (const archive of ARCHIVES) {
    const destDir = join(CACHE, archive.name, archive.version);
    if (existsSync(destDir) && existsSync(join(CACHE, archive.name, `${archive.version}.7z`))) {
      console.log(`✓ ${archive.name} cache ready (${destDir})`);
      continue;
    }

    console.log(`\n=== Preparing ${archive.name} cache ===`);

    // Reset target dir
    rmSync(destDir, { recursive: true, force: true });
    mkdirSync(destDir, { recursive: true });

    // Download to temp, then extract
    const zipPath = join(tmpdir(), `${archive.name}-${Date.now()}.7z`);
    try {
      await download(archive.url, zipPath);
      // Also save a copy to the cache dir so electron-builder recognizes it
      const cacheZip = join(CACHE, archive.name, `${archive.version}.7z`);
      require('fs').copyFileSync(zipPath, cacheZip);
      extract(zipPath, destDir, archive.needsSnl);
      console.log(`  ✓ ${archive.name} ready`);
    } finally {
      try { rmSync(zipPath, { force: true }); } catch { /* ignore */ }
    }
  }

  console.log('\n✓ All electron-builder caches ready');
}

main().catch(err => {
  console.error('Failed:', err.message);
  process.exit(1);
});

File: scripts/prepare-electron-builder-cache.py
#!/usr/bin/env python3
"""Pre-populate electron-builder caches for Windows release builds.
Downloads winCodeSign, NSIS, and NSIS-resources archives from
electron-builder-binaries releases and extracts them (skipping
symlinks where needed)."""

import urllib.request, os, subprocess, shutil, ssl, sys

CACHE = os.path.join(os.environ['LOCALAPPDATA'], 'electron-builder', 'Cache')
SEVEN_ZIP = os.path.join('node_modules', '7zip-bin', 'win', 'x64', '7za.exe')

ARCHIVES = [
    # Name, version, URL, skip-symlinks
    ('winCodeSign', 'winCodeSign-2.6.0',
     'winCodeSign-2.6.0/winCodeSign-2.6.0.7z', True),
    ('nsis', 'nsis-3.0.4.1',
     'nsis-3.0.4.1/nsis-3.0.4.1.7z', False),
    ('nsis-resources', 'nsis-3.0.4.1',
     'nsis-resources-3.4.1/nsis-resources-3.4.1.7z', False, 'nsis'),
]

BASE_URL = 'https://github.com/electron-userland/electron-builder-binaries/releases/download'

ctx = ssl.create_default_context()
ctx.check_hostname = False
ctx.verify_mode = ssl.CERT_NONE
os.makedirs(CACHE, exist_ok=True)

for args in ARCHIVES:
    name, ver, rel_path, needs_snl = args[:4]
    cache_name = name
    dest = os.path.join(CACHE, name, ver)
    cache_zip = os.path.join(CACHE, name, ver + '.7z')
    if os.path.isdir(dest) and os.path.exists(cache_zip):
        print(f'~ {name} cache ready')
        continue

    print(f'\n=== Preparing {name} cache ===')
    if os.path.isdir(dest):
        shutil.rmtree(dest)
    os.makedirs(dest)

    url = f'{BASE_URL}/{rel_path}'
    zip_path = os.path.join(os.environ.get('TEMP', r'C:\tmp'), f'{name}-cache.7z')
    try:
        print(f'  downloading {url}')
        req = urllib.request.Request(url, headers={'User-Agent': 'node'})
        with urllib.request.urlopen(req, context=ctx, timeout=30) as resp:
            with open(zip_path, 'wb') as f:
                f.write(resp.read())
        print(f'  saved {os.path.getsize(zip_path) / 1048576:.1f} MiB')

        # Save clean copy for electron-builder to recognize
        shutil.copy2(zip_path, cache_zip)

        args = [SEVEN_ZIP, 'x', zip_path, f'-o{dest}', '-y']
        if needs_snl:
            args.append('-snl')
        r = subprocess.run(args)
        if r.returncode != 0:
            print(f'  7za extract failed (exit {r.returncode})')
            sys.exit(1)
        print(f'  ~ {name} ready')
    except Exception as e:
        print(f'Failed: {e}')
        sys.exit(1)
    finally:
        try: os.remove(zip_path)
        except: pass

print('\n~ All caches ready')
